### ISSUE TITLE
Access is denied when fetching artifacts

### ISSUE THREAD

**Issue by reidca** [gocd/gocd#1428]
When fetching artifacts in some pipelines we see "Access is denied" messages.

These failures appear in certain pipelines repeatedly however the exact same pipelines/stage/job can also execute OK.

I have tried the usual combination of re-running the pipeline and its dependences from start, restarting the GO Server, the GO Agent in question however the problem persists.

Here is an example:
```
0:18:50.632 [go] Start to execute task: <fetchartifact pipeline="MyPipeline" stage="stage_package" job="job_package" srcdir="Scripts" />.
10:18:50.632 [go] Fetching artifact [Scripts] from [MyPipeline/334/stage_package/1/job_package]
10:18:55.694 [go] Could not fetch artifact https://MyServer:443/go/remoting/files/MyPipeline/334/stage_package/1/job_package/Scripts.zip. Pausing 20 seconds to retry. Error was : Caught an exception 'pipelines\WHSLE_Web_Assurance_Deploy_Testing\Scripts\DeployTools\Deploy-Modules.ps1 (Access is denied)'
10:19:15.726 [go] Could not fetch artifact https://MyServer:443/go/remoting/files/MyPipeline/334/stage_package/1/job_package/Scripts.zip. Pausing 29 seconds to retry. Error was : Caught an exception 'pipelines\WHSLE_Web_Assurance_Deploy_Testing\Scripts\DeployTools\Deploy-Modules.ps1 (Access is denied)'
10:19:44.742 [go] Could not fetch artifact https://MyServer:443/go/remoting/files/MyPipeline/334/stage_package/1/job_package/Scripts.zip. Pausing 39 seconds to retry. Error was : Caught an exception 'pipelines\WHSLE_Web_Assurance_Deploy_Testing\Scripts\DeployTools\Deploy-Modules.ps1 (Access is denied)'
10:20:23.758 [go] Failed to save artifact [Scripts] to []
10:20:23.789 [go] Current job status: failed.
```

The corresponding error in on the agent is:
```
2015-08-10 11:43:56,643 [loopThread] INFO  thoughtworks.go.work.DefaultGoPublisher:95 - Agent [sd-wh-web1-dev, 10.129.10.150, 4e94fe64-6264-4337-9e0e-faccbdd043b7] is reporting status [Building] to Go Server for Build [WHSLE_Web_Assurance_Deploy_Testing/173/Deploy_API/2/Deploy_API/95271]
2015-08-10 11:43:56,706 [loopThread] INFO  thoughtworks.go.util.HttpService:129 - Got back 200 from server
2015-08-10 11:43:56,706 [loopThread] INFO  thoughtworks.go.util.PerfTimer:53 - Performance: Downloading from url [https://MyServer:443/go/remoting/files/WHSLE_PubFacing_AssuranceAPI_Commit/334/stage_package/1/job_package/cruise-output/md5.checksum] took 47ms
2015-08-10 11:43:56,722 [loopThread] INFO  thoughtworks.go.domain.ChecksumFileHandler:58 - [Agent Fetch Artifact] Saved checksum property file [pipelines\WHSLE_Web_Assurance_Deploy_Testing\Packages\Testing\WHSLE_PubFacing_AssuranceAPI_Commit_stage_package_job_package_md5.checksum]
2015-08-10 11:43:57,378 [loopThread] INFO  thoughtworks.go.util.HttpService:129 - Got back 304 from server
2015-08-10 11:43:57,378 [loopThread] INFO  thoughtworks.go.util.PerfTimer:53 - Performance: Downloading from url [https://MyServer:443/go/remoting/files/WHSLE_PubFacing_AssuranceAPI_Commit/334/stage_package/1/job_package/Packages/Testing/WholesaleAssuranceAPI.zip?sha1=Ny2bEREzkmJh7rqTXudQGIAs9%2Bk%3D] took 125ms
2015-08-10 11:43:57,393 [loopThread] INFO  thoughtworks.go.util.HttpService:129 - Got back 200 from server
2015-08-10 11:43:57,393 [loopThread] INFO  thoughtworks.go.util.PerfTimer:53 - Performance: Downloading from url [https://MyServer:443/go/remoting/files/WHSLE_PubFacing_AssuranceAPI_Commit/334/stage_package/1/job_package/cruise-output/md5.checksum] took 15ms
2015-08-10 11:43:57,393 [loopThread] INFO  thoughtworks.go.domain.ChecksumFileHandler:58 - [Agent Fetch Artifact] Saved checksum property file [pipelines\WHSLE_Web_Assurance_Deploy_Testing\WHSLE_PubFacing_AssuranceAPI_Commit_stage_package_job_package_md5.checksum]
2015-08-10 11:43:57,425 [loopThread] INFO  thoughtworks.go.util.HttpService:129 - Got back 200 from server
2015-08-10 11:43:57,425 [loopThread] INFO  thoughtworks.go.util.PerfTimer:53 - Performance: Downloading from url [https://MyServer:443/go/remoting/files/WHSLE_PubFacing_AssuranceAPI_Commit/334/stage_package/1/job_package/Scripts.zip] took 32ms
2015-08-10 11:43:57,425 [loopThread] INFO  thoughtworks.go.domain.DirHandler:53 - [Agent Fetch Artifact] Downloading from 'Scripts' to 'C:\Program Files (x86)\Go Agent\pipelines\WHSLE_Web_Assurance_Deploy_Testing'. Will read from Socket stream to compute MD5 and write to file
2015-08-10 11:43:57,425 [loopThread] ERROR thoughtworks.go.util.ZipUtil:158 - Failed to unzip file [Scripts/DeployTools/Deploy-Modules.ps1] to directory [C:\Program Files (x86)\Go Agent\pipelines\WHSLE_Web_Assurance_Deploy_Testing]
java.io.FileNotFoundException: pipelines\WHSLE_Web_Assurance_Deploy_Testing\Scripts\DeployTools\Deploy-Modules.ps1 (Access is denied)
	at java.io.FileOutputStream.open(Native Method)
	at java.io.FileOutputStream.<init>(Unknown Source)
	at java.io.FileOutputStream.<init>(Unknown Source)
	at com.thoughtworks.go.util.ZipUtil.extractTo(ZipUtil.java:140)
	at com.thoughtworks.go.util.ZipUtil.unzip(ZipUtil.java:119)
	at com.thoughtworks.go.domain.DirHandler.handle(DirHandler.java:56)
	at com.thoughtworks.go.util.HttpService.download(HttpService.java:101)
	at com.thoughtworks.go.domain.DownloadAction.download(DownloadAction.java:73)
	at com.thoughtworks.go.domain.DownloadAction.perform(DownloadAction.java:47)
	at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.pullArtifact(FetchArtifactBuilder.java:73)
	at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.downloadArtifact(FetchArtifactBuilder.java:64)
	at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.fetch(FetchArtifactBuilder.java:59)
	at com.thoughtworks.go.publishers.GoArtifactsManipulator.fetch(GoArtifactsManipulator.java:183)
	at com.thoughtworks.go.work.DefaultGoPublisher.fetch(DefaultGoPublisher.java:77)
	at com.thoughtworks.go.domain.builder.FetchArtifactBuilder.build(FetchArtifactBuilder.java:54)
	at com.thoughtworks.go.domain.builder.Builder.build(Builder.java:51)
	at com.thoughtworks.go.remote.work.Builders.build(Builders.java:63)
	at com.thoughtworks.go.remote.work.BuildWork.execute(BuildWork.java:208)
	at com.thoughtworks.go.remote.work.BuildWork.buildJob(BuildWork.java:183)
	at com.thoughtworks.go.remote.work.BuildWork.build(BuildWork.java:145)
	at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:91)
	at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:53)
	at com.thoughtworks.go.agent.AgentController.retrieveWork(AgentController.java:186)
	at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:132)
	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273)
	at org.springframework.scheduling.support.MethodInvokingRunnable.run(MethodInvokingRunnable.java:65)
	at org.springframework.scheduling.timer.DelegatingTimerTask.run(DelegatingTimerTask.java:70)
	at java.util.TimerThread.mainLoop(Unknown Source)
	at java.util.TimerThread.run(Unknown Source)
```

If I look into the directory I can see that the files have all be fetched successfully however they are all read-only.

Looking at other pipelines and jobs that fetch the same files they are not read-only and those jobs work OK. I have checked the files on the job where they are uploaded, the same files in the artifact directory and they are not read-only anywhere.

Using Process Monitor I can see that the files get set to be read-only and then a call to overwrite the files occurs however fails with ACCESS IS DENIED resulting in this error. This is failing even though the files are actually there. I have no idea why it is trying to overwrite the files when they have already been written.
See screen shots
![go access denied 1](https://cloud.githubusercontent.com/assets/7325010/9170256/2b8343d4-3f59-11e5-8e8e-33009f0b509a.png)
![go access denied 2](https://cloud.githubusercontent.com/assets/7325010/9170262/34dae996-3f59-11e5-849d-1cb5bcc59dc0.png)

Looking at the [source code](https://github.com/gocd/gocd/blob/master/base/src/com/thoughtworks/go/util/ZipUtil.java) it appears that this problem occurs in the "extractTo" function when calling [IOUtils.copyLarge(entryInputStream, os);](https://commons.apache.org/proper/commons-io/apidocs/src-html/org/apache/commons/io/IOUtils.html) however the source code for this does not reveal anything to me.

This problem is causing a serious issue for us. People in our organisation are starting to lose confidence in the platform and I would appreciate anyone being able to provide assistance, work-around or possible fixes.

**Comment by ketan** [comment]
`IOUtils.copyLarge(entryInputStream, os)` just copies the bytes from one stream into another in an efficient manner, it does not do the file creation.

Looking at the code that [handles zipping](https://github.com/gocd/gocd/blob/master/base/src/com/thoughtworks/go/util/ZipUtil.java#L99-L102) and [handles extraction](https://github.com/gocd/gocd/blob/master/base/src/com/thoughtworks/go/util/ZipUtil.java#L135-L170), we don't preserve file "mode" or "permissions" on any of the files/directories during compression/decompression.

There are a few possibilities that come to mind that could possibly explain this odd behaviour, these are just hypotheticals, and some or all of them may not apply. I'm just trying to see what specific explanations we can rule out at this point.

* The permission model on windows when the go agent runs "as a service account" vs "running as a user".
* `C:\Program Files (x86)` is a treated as a protected directory of sorts by windows, so we suspect that this could be what forces the file/directory mode to be read only
* The build was run by a different user from the same directory as go runs, causing the file permissions to be messed up (either via RDP) or someone running `agent.bat` or `agent.cmd` as a different user account.
* The build probably ran something by changing users using [runas](https://technet.microsoft.com/en-us/library/cc771525.aspx) or some similar mechanism to change user account

In order to diagnose further, can you please check the following -

* what user is running the `java` process under question
* which user/group owns the file under question
* which user/group owns each of the directories leading upto that file (is there a trace available when these dirs are created)
* why is the `CreateFile` call invoked so many times for the same file, we're not sure why this is happening, is there a way to export multiple events from procmon?

Here are some possible "workarounds" that we have to offer -

* Set user permissions on the `java.exe` to allow all access (from [stackoverflow](http://stackoverflow.com/questions/19561386/access-is-denied-java-io-filenotfoundexception/28689965#28689965)) - this is a dangerous operation, but may be useful to diagnose the problem.
* Run go agent as a normal user by clicking "agent.cmd" or "agent.bat", by installing the agent from a [zipfile](http://download.go.cd/gocd/go-agent-15.2.0-2248.zip) extracted to `c:\go-agent`

**Comment by reidca** [comment]
Thanks for looking at this. I really hope to get a resolution as this is a very frustrating issue for us.

The installation of GO Agent on this server was performed using the installation wizard. Normally we change the install directory to c:\GOAgent (for reasons of keeping the path length shorter really) however this is an old installation and is therefore using the defaults. 

We have not seen any problems using the "program files" path before although I have previously submitted issue #484 to make this configurable as the pipelines are really "data" and "program files" is for content that does not change,

> what user is running the java process under question

The user running Java.exe is "NT Authority\System" because the service is running as LOCAL SYSTEM (again as per the defaults).
![go access denied 3](https://cloud.githubusercontent.com/assets/7325010/9175248/92559760-3f7c-11e5-9a77-caaa67949fde.png)

> which user/group owns the file under question

Files is owned by Administrators however LOCAL SYSTEM has FULL CONTROL to all files on under "c:\Program Files (x86)" 

![go access denied 4](https://cloud.githubusercontent.com/assets/7325010/9175384/5635e702-3f7d-11e5-8c52-28bb72cb1796.png)

> which user/group owns each of the directories leading upto that file (is there a trace available when these dirs are created)

NTFS permissions are being inherited from root. LOCAL SYSTEM is a very powerful user in Windows and has full control of all directories. You can see in this picture that it has full control of the containing folder and the file in question.

![go access denied 5](https://cloud.githubusercontent.com/assets/7325010/9175385/5ac7be9e-3f7d-11e5-9b4e-99be9fa6adec.png)

![go access denied 6](https://cloud.githubusercontent.com/assets/7325010/9175458/bb121358-3f7d-11e5-9617-91c8046e1127.png)

> why is the CreateFile call invoked so many times for the same file, we're not sure why this is happening, is there a way to export multiple events from procmon?

This I do not know. There is clearly something in the Java interaction with Windows environment that is causing this to happen. The CreateFile command is a Win32 API call and it gets called even if the process only wants to read the file. I am more interested in why the Readonly attribute is being set and then why it tries to write to the file. If readonly was not being set then the access denied would not occur.

I have the put the PML file which shows these events [here](http://1drv.ms/1PgJrIA). You can open these in Process Monitor.

**Comment by reidca** [comment]
Small update, to rule out the "Program Files" directory issue I just reinstalled the GO Agent (15.1) to c:\GoAgent and ran the pipeline again. I have found the exact same error, the files are again read-only and "Access is denied" error is recorded in logs.

**Comment by zabil** [comment]
@reidca I assume this was done using the installer did you try what @ketan mentioned? 
> Run go agent as a normal user by clicking "agent.cmd" or "agent.bat", by installing the agent from a zipfile extracted to c:\go-agent

**Comment by reidca** [comment]
> @reidca I assume this was done using the installer did you try what @ketan mentioned?

I have just tried this now with the exact same result.

Here is what I did:

1. Stopped the existing GO Agent service. 
2. Downloaded the [Java zip version of the agent](http://download.go.cd/gocd/go-agent-15.1.0-1863.zip)
3. Checked the environment variable GO_AGENT_JAVA_HOME were pointing to the JRE from the existing agent install
4. Added Java bin directory to path
5. Opened an elevated command prompt as a Windows domain user with local admin rights
6. Ran Start-Agent.bat
7. Enabled the agent
8. Assigned the appropriate environments and resources to the agent

The job picked up and executing but the result was identical, read-only files created and "Access is denied" message in the console.

Any other ideas? This is driving me crazy.

**Comment by ketan** [comment]
@reidca -

Another possibility that comes to mind - Files of type `.ps1` are powershell scripts. Untrusted scripts download from elsewhere may not have the right permission levels as an OS policy.

To eliminate this possibility -

* Can you rename the `.ps1` files to say `.txt` files and try again, preferably using the zipfile installer running from `c:\go-agent`
* As a fallback (dangerous), set the [execution policy](https://technet.microsoft.com/en-us/library/ee176961.aspx) to `Unrestricted`, make sure to restore the execution policy back if this fails

**Comment by reidca** [comment]
@ketan Thanks for the suggestion however we use these Powershell scripts in ALL our pipelines and this problem does not happen elsewhere. In fact the very same directory "scripts" fetches without any problem on all the other pipelines at the moment. Although in the past I have seen similar errors on other agents that happen at random.

**Comment by reidca** [comment]
This issue has now been resolved, it turned out to be something very obvious with hindsight, which is often the case I find.

The problem started on this pipeline after some rationalisation of script folder names across all our pipelines. When we use scripts, previously the name of the folder that the scripts were being downloaded to was not always the same, even if the scripts were the same and coming from the same location in source control.

This was recently bulk-updated across all our pipelines to ensure all artifacts and fetch commands that used scripts used the same folder name and artifact name.

Shortly after this was done the problem started to happen. However I did not consider the problem from a pipeline design perspective, I instead focused on issued on the server or the agent communication.

The problem turned out to be that "Fetch Materials" was erroneously specified on the stage settings, the "Scripts" artifact from the parent pipeline was being fetched into a folder called "Scripts" in a task, and at the same time there was a source control material specified in the pipeline which was also downloaded into a folder called "Scripts". I assume that the "Fetch Materials" occurred first, created the "Scripts" folder with read-only files, then the "Fetch Artifact" task ran and could not write the files as they were read-only.

The reason the problem did not occur previously was because of the difference in folder names. The "Fetch Materials" command was using a folder called "Tools" whilst the "Fetch Artifact" was using "Scripts\Tools". As soon as these were consolidated the problem occurred however it was not obvious as the update was done in bulk from the XML.

The error message did not really help resolve the issue, especially as it looks like an issue fetching the materials from the GO server over HTTP, however the problem was not a technical issues with GO. I will submit a separate issue on improving the error message.

Sorry for wasting everyone's time on this, your efforts were appreciated.
